### ISSUE TITLE
Another round of failing tests 

### ISSUE THREAD

**Issue by braxtonhall** [jyoo980/silver#38]
Please don't look at this like a checklist. I'm just wondering if there's a fundamental change due to inlining that we can talk about that causes a bunch of these. I assume this is true of internal-bug-2, test7 and unfolding.
# 
[`internal-bug-2`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/chalice/internal-bug-2.vpr)
```
2 errors

The following outputs were expected according to the test annotations, but did not occur during testing:
  assignment.failed:insufficient.permission (internal-bug-2.vpr:12)

The following output occurred during testing, but should not have according to the test annotations:
  [assert.failed:assertion.false] [assert.failed:assertion.false] Assert might fail. Assertion this.x == old(this.x) might not hold. (internal-bug-2.vpr@14.3)
```
# 
[`test7`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/chalice/test7.vpr)
I _think_ this is the same as fold in #34 
```
5 errors

The following outputs were expected according to the test annotations, but did not occur during testing:
  not.wellformed:insufficient.permission (test7.vpr:112)
  application.precondition:insufficient.permission (test7.vpr:106)
  assert.failed:insufficient.permission (test7.vpr:89)
  assert.failed:insufficient.permission (test7.vpr:67)
```
# 
[`unfolding.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/functions/unfolding.vpr)
Looks like a duplicate of internal-bug-2
```
1 errors

The following output occurred during testing, but should not have according to the test annotations:
  [assignment.failed:insufficient.permission] [assignment.failed:insufficient.permission] Assignment might fail. There might be insufficient permission to access pair(y) (unfolding.vpr@75.3)
```
# 
[`triggerFoldPackage.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/heap-dependent_triggers/triggerFoldPackage.vpr)
This one has especially spooky output
```
1 errors

The following output occurred during testing, but should not have according to the test annotations:
  [exceptional.error] Verification aborted exceptionally (<no position>)
```
# 
# 
And then there are a bunch of the None.get
```
None.get
java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:349)
	at scala.None$.get(Option.scala:347)
	at viper.silver.plugin.standard.inline.InlineRewrite$$anonfun$viper$silver$plugin$standard$inline$InlineRewrite$$expandExpression$1.applyOrElse(InlineRewrite.scala:44)
	at viper.silver.plugin.standard.inline.InlineRewrite$$anonfun$viper$silver$plugin$standard$inline$InlineRewrite$$expandExpression$1.applyOrElse(InlineRewrite.scala:40)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:34)
        ...
```
These occur in a lot of tests like [this one](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/basic/many_conjuncts.vpr)

**Comment by jyoo980** [comment]
Some (if not most) of the `None.get` errors seem to be resolved with #39

**Comment by braxtonhall** [comment]
The None errors have all be resolved, however all the other tests are still failing (even with #40)

**Comment by ionathanch** [comment]
(Marking things to look at or talk about in the writeup with checkboxes so I don't forget.)

# Spooky Scary Errors

[`triggerFoldPackage`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/heap-dependent_triggers/triggerFoldPackage.vpr) is still spooky

- [x] [`carbon/0196.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0196.vpr) has an unusual error message. ~I think it has to do with mutual recursion detection in #5, where we use a call graph.~ Fixed with #42

## Java errors for `Unfolding`

[`silicon/0092.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0092.vpr):
```
[info] - all/issues/silicon/0092.vpr [Silicon-Silver] *** FAILED *** (58 milliseconds)
[info]   java.lang.IllegalArgumentException: argument type mismatch
[info]   at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
[info]   at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
[info]   at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[info]   at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
[info]   at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaMethodMirror.jinvokeraw(JavaMirrors.scala:339)
[info]   at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaMethodMirror.jinvoke(JavaMirrors.scala:342)
[info]   at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaVanillaMethodMirror.apply(JavaMirrors.scala:358)
[info]   at viper.silver.ast.utility.rewriter.Rewritable.withChildren(Rewritable.scala:59)
[info]   at viper.silver.ast.utility.rewriter.Rewritable.withChildren$(Rewritable.scala:24)
[info]   at viper.silver.ast.Unfolding.withChildren(Expression.scala:454)
```

[`silicon/0290.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0290.vpr):
```
[info] - all/issues/silicon/0290.vpr [Silicon-Silver] *** FAILED *** (44 milliseconds)
[info]   java.lang.IllegalArgumentException: java.lang.ClassCastException@6e0a2e20
[info]   at jdk.internal.reflect.GeneratedConstructorAccessor52.newInstance(Unknown Source)
[info]   at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[info]   at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
[info]   at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaMethodMirror.jinvokeraw(JavaMirrors.scala:339)
[info]   at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaMethodMirror.jinvoke(JavaMirrors.scala:342)
[info]   at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaVanillaMethodMirror.apply(JavaMirrors.scala:358)
[info]   at viper.silver.ast.utility.rewriter.Rewritable.withChildren(Rewritable.scala:59)
[info]   at viper.silver.ast.utility.rewriter.Rewritable.withChildren$(Rewritable.scala:24)
[info]   at viper.silver.ast.Unfolding.withChildren(Expression.scala:454)
[info]   at viper.silver.ast.utility.rewriter.Strategy.rewriteTopDown(Strategy.scala:410)
```

[`silicon/0292.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0292.vpr):
```
[info] - all/issues/silicon/0292.vpr [Silicon-Silver] *** FAILED *** (38 milliseconds)
[info]   java.lang.IllegalArgumentException: java.lang.ClassCastException@78e8a26a
[info]   at jdk.internal.reflect.GeneratedConstructorAccessor52.newInstance(Unknown Source)
[info]   at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
[info]   at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
[info]   at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaMethodMirror.jinvokeraw(JavaMirrors.scala:339)
[info]   at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaMethodMirror.jinvoke(JavaMirrors.scala:342)
[info]   at scala.reflect.runtime.JavaMirrors$JavaMirror$JavaVanillaMethodMirror.apply(JavaMirrors.scala:358)
[info]   at viper.silver.ast.utility.rewriter.Rewritable.withChildren(Rewritable.scala:59)
[info]   at viper.silver.ast.utility.rewriter.Rewritable.withChildren$(Rewritable.scala:24)
[info]   at viper.silver.ast.Unfolding.withChildren(Expression.scala:454)
[info]   at viper.silver.ast.utility.rewriter.Strategy.rewriteTopDown(Strategy.scala:410)
```

# (Mostly) Understood Errors

## No more error/Different error

As mentioned in #40, the original first three now fail with different messages and seem to be about the same issue:
- [x] Failing folds/unfolds due to insufficient permissions, or failing permissions assertions (e.g. 0081.vpr) no longer fail, since inlining provides the previously-missing permissions.
- [x] ~[`carbon/0060.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0060.vpr) is interesting: It asserts the current predicate and field access permissions at different stages of unfolding.~ Fixed by #44

[`silicon/0183.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0183.vpr) is interesting. Without removing the unfolding, there's a nonlinear permission constraint that Z3 can't solve and throws an error, so this is a bug in upstream Viper, but when we remove the unfolding, the nonlinear permission issue disappears, and there's no more error.

A few more failing tests that look to be about the same no-fold/unfolds failing-to-fail:
* [`carbon/0042.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0042.vpr)
* [`carbon/0049.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0049.vpr)
* [`carbon/0053.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0053.vpr)
* [`carbon/0063.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0063.vpr)
* [`carbon/0080.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0080.vpr)
* [`carbon/0081.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0081.vpr)
* [`carbon/0125.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0125.vpr)
* [`carbon/0150.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0150.vpr)
* [`carbon/0213.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0213.vpr)
* [`carbon/0271.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0271.vpr)
* [`silicon/0232.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0232.vpr)

## Unimplemented features

[`triggerPred.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/heap-dependent_triggers/triggerPred.vpr) is interesting:
```
[info]   3 errors
[info]   
[info]   The following output occurred during testing, but should not have according to the test annotations:
[info]     [assert.failed:assertion.false] [assert.failed:assertion.false] Assert might fail. Assertion perm(p(y)) == write might not hold. (triggerPred.vpr@27.5)
[info]     [assert.failed:insufficient.permission] [assert.failed:insufficient.permission] Assert might fail. There might be insufficient permission to access y.f (triggerPred.vpr@17.5)
[info]     [assert.failed:insufficient.permission] [assert.failed:insufficient.permission] Assert might fail. There might be insufficient permission to access y.f (triggerPred.vpr@10.5) (AnnotationBasedTestSuite.scala:63)
```
- [x] Looks like it's due to quantified expressions and the interactions between triggers and inlined predicates.
Others due to quantified expression:
* [`carbon/0237.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/carbon/0237.vpr)
* [`silicon/0114.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0114.vpr)
* [`silver/0227.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0227.vpr)
* [`silver/0228c.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0228c.vpr)
* [`silver/0240.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0240.vpr)
* [`silver/0252.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0252.vpr)
* [`silver/0336b.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0336b.vpr)
* [`silver/0383a.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0383a.vpr)
* [`silver/0383b.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0383b.vpr)
* [`silver/0393.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0393b.vpr)
* [`silver/0493a.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0493a.vpr)
* [`silver/0493b.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0493b.vpr)

[`silver/0135.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0135.vpr) is due to domains, which we know we haven't implemented.

These ones are all due to magic wands:
* [`silver/0213.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0213.vpr)
* [`silver/0221.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0221.vpr)
* [`silver/0229.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0229.vpr)

# To Look At

A few I didn't expect to fail:
* [`silicon/0119.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0119.vpr)
* [`silicon/0379b.vpr`](https://github.com/jyoo980/silver/blob/master/src/test/resources/all/issues/silicon/0379b.vpr)

After `silver/0062.vpr` the tests started to hang, so I stopped them there. I'm not sure how many tests were actually run in total, since I ran this on the terminal.

**Comment by jyoo980** [comment]
@ionathanch 

> **carbon/0196.vpr has an unusual error message:**

The only interesting I see here is that the predicate `Q` doesn't have a body. Maybe constructing the predicate call graph with _all_ predicates (regardless of whether they have a body or not) will fix this.
